Hash discrepancy with stdlib('c') use

I've run into a strange issue in https://github.com/bioconda/bioconda-recipes/pull/64176 where adding `stdlib('c')` to the build section causes the package build to fail due to a difference in the hashes, as seen in this section of a failing log

```
2026-04-07T14:17:57.0723124Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0723558Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) Resource usage statistics from testing nohuman:[0m
2026-04-07T14:17:57.0724032Z 14:17:44 [32mBIOCONDA INFO[0m (OUT)    Process count: 2[0m
2026-04-07T14:17:57.0724490Z 14:17:44 [32mBIOCONDA INFO[0m (OUT)    CPU time: Sys=-, User=0:00:00.0[0m
2026-04-07T14:17:57.0724929Z 14:17:44 [32mBIOCONDA INFO[0m (OUT)    Memory: 14.1M[0m
2026-04-07T14:17:57.0725300Z 14:17:44 [32mBIOCONDA INFO[0m (OUT)    Disk usage: 16B[0m
2026-04-07T14:17:57.0725703Z 14:17:44 [32mBIOCONDA INFO[0m (OUT)    Time elapsed: 0:00:02.0[0m
2026-04-07T14:17:57.0726078Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0726367Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0726895Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) TEST END: /opt/conda/conda-bld/linux-64/nohuman-0.5.1-h4bf21ff_0.conda[0m
2026-04-07T14:17:57.0728057Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) Renaming work directory '/opt/conda/conda-bld/nohuman_1775571179039/work' to '/opt/conda/conda-bld/nohuman_1775571179039/work_moved_nohuman-0.5.1-hee99bd0_0_linux-64_main_build_loop'[0m
2026-04-07T14:17:57.0729517Z 14:17:44 [32mBIOCONDA INFO[0m (OUT) shutil.move(work)=/opt/conda/conda-bld/nohuman_1775571179039/work, dest=/opt/conda/conda-bld/nohuman_1775571179039/work_moved_nohuman-0.5.1-hee99bd0_0_linux-64_main_build_loop)[0m
2026-04-07T14:17:57.0730460Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) # Automatic uploading is disabled[0m
2026-04-07T14:17:57.0731302Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) # If you want to upload package(s) to anaconda.org later, type:[0m
2026-04-07T14:17:57.0731771Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0732063Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0732538Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) # To have conda build upload to anaconda.org automatically, use[0m
2026-04-07T14:17:57.0733355Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) # conda config --set anaconda_upload yes[0m
2026-04-07T14:17:57.0733809Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) anaconda upload \[0m
2026-04-07T14:17:57.0734377Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)     /opt/conda/conda-bld/linux-64/nohuman-0.5.1-h4bf21ff_0.conda[0m
2026-04-07T14:17:57.0735004Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) anaconda_upload is not set.  Not uploading wheels: [][0m
2026-04-07T14:17:57.0735446Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0736110Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) INFO :: The inputs making up the hashes for the built packages are as follows:[0m
2026-04-07T14:17:57.0736615Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) {[0m
2026-04-07T14:17:57.0737021Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)   "nohuman-0.5.1-h4bf21ff_0.conda": {[0m
2026-04-07T14:17:57.0737452Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)     "recipe": {[0m
2026-04-07T14:17:57.0737878Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "CONDA_BUILD_SYSROOT": "",[0m
2026-04-07T14:17:57.0738353Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "__glibc": "__glibc >=2.17,<3.0.a0",[0m
2026-04-07T14:17:57.0738813Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "c_compiler": "gcc",[0m
2026-04-07T14:17:57.0739282Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "c_compiler_version": "14",[0m
2026-04-07T14:17:57.0739730Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "c_stdlib": "sysroot",[0m
2026-04-07T14:17:57.0740161Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "c_stdlib_version": "2.17",[0m
2026-04-07T14:17:57.0741032Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "channel_targets": "bioconda main",[0m
2026-04-07T14:17:57.0741715Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "rust_compiler": "rust",[0m
2026-04-07T14:17:57.0742185Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "target_platform": "linux-64",[0m
2026-04-07T14:17:57.0742602Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)       "xz": "5"[0m
2026-04-07T14:17:57.0742934Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)     }[0m
2026-04-07T14:17:57.0743255Z 14:17:45 [32mBIOCONDA INFO[0m (OUT)   }[0m
2026-04-07T14:17:57.0743546Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) }[0m
2026-04-07T14:17:57.0743837Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0744120Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0744632Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) ####################################################################################[0m
2026-04-07T14:17:57.0745138Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) Resource usage summary:[0m
2026-04-07T14:17:57.0745493Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) [0m
2026-04-07T14:17:57.0745837Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) Total time: 0:04:46.6[0m
2026-04-07T14:17:57.0746281Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) CPU usage: sys=0:00:09.3, user=0:03:16.5[0m
2026-04-07T14:17:57.0746779Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) Maximum memory usage observed: 1.0G[0m
2026-04-07T14:17:57.0747332Z 14:17:45 [32mBIOCONDA INFO[0m (OUT) Total disk usage observed (not including envs): 150.0M[0m
2026-04-07T14:17:57.0747924Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 Subdir: linux-64 Gathering repodata[0m
2026-04-07T14:17:57.0748499Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-64 Writing pre-patch repodata[0m
2026-04-07T14:17:57.0749073Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-64 Applying patch instructions[0m
2026-04-07T14:17:57.0749649Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-64 Writing patched repodata[0m
2026-04-07T14:17:57.0750246Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-64 Building current_repodata subset[0m
2026-04-07T14:17:57.0751116Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 Subdir: linux-aarch64 Gathering repodata[0m
2026-04-07T14:17:57.0751738Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-aarch64 Writing pre-patch repodata[0m
2026-04-07T14:17:57.0752355Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-aarch64 Applying patch instructions[0m
2026-04-07T14:17:57.0753142Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-aarch64 Writing patched repodata[0m
2026-04-07T14:17:57.0753723Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 Subdir: noarch Gathering repodata[0m
2026-04-07T14:17:57.0754341Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-aarch64 Building current_repodata subset[0m
2026-04-07T14:17:57.0754953Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 noarch Writing pre-patch repodata[0m
2026-04-07T14:17:57.0755517Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 noarch Applying patch instructions[0m
2026-04-07T14:17:57.0756083Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 noarch Writing patched repodata[0m
2026-04-07T14:17:57.0756673Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 noarch Building current_repodata subset[0m
2026-04-07T14:17:57.0757273Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-64 Writing current_repodata subset[0m
2026-04-07T14:17:57.0757852Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-64 Writing index HTML[0m
2026-04-07T14:17:57.0758425Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 noarch Writing current_repodata subset[0m
2026-04-07T14:17:57.0759051Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-aarch64 Writing current_repodata subset[0m
2026-04-07T14:17:57.0759803Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 noarch Writing index HTML[0m
2026-04-07T14:17:57.0760375Z 14:17:55 [32mBIOCONDA INFO[0m (OUT) 2026-04-07T14:17:55 linux-aarch64 Writing index HTML[0m
2026-04-07T14:17:57.0760794Z 
2026-04-07T14:17:57.0761387Z 14:17:57 [31mBIOCONDA ERROR[0m BUILD FAILED: the built package /opt/mambaforge/envs/bioconda/conda-bld/linux-64/nohuman-0.5.1-hee99bd0_0.conda cannot be found[0m
2026-04-07T14:17:57.0762365Z 14:17:57 [32mBIOCONDA INFO[0m Finished build for recipes/nohuman Free disk space: 116857.66 MB. Free memory: 10943.81 MB (68.44%)[0m
2026-04-07T14:17:57.0762977Z 14:17:57 [32mBIOCONDA INFO[0m (COMMAND) docker system df[0m
2026-04-07T14:17:57.0763478Z 14:17:57 [32mBIOCONDA INFO[0m (OUT) TYPE            TOTAL     ACTIVE    SIZE      RECLAIMABLE[0m
2026-04-07T14:17:57.0764050Z 14:17:57 [32mBIOCONDA INFO[0m (OUT) Images          1         0         2.061GB   2.061GB (100%)[0m
2026-04-07T14:17:57.0764588Z 14:17:57 [32mBIOCONDA INFO[0m (OUT) Containers      0         0         0B        0B[0m
2026-04-07T14:17:57.0765123Z 14:17:57 [32mBIOCONDA INFO[0m (OUT) Local Volumes   0         0         0B        0B[0m
2026-04-07T14:17:57.0765630Z 14:17:57 [32mBIOCONDA INFO[0m (OUT) Build Cache     0         0         0B        0B[0m
2026-04-07T14:17:57.0766067Z 14:17:57 [32mBIOCONDA INFO[0m (COMMAND) docker ps -a[0m
2026-04-07T14:17:57.0766606Z 14:17:57 [32mBIOCONDA INFO[0m (OUT) CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES[0m
2026-04-07T14:17:57.0767138Z 14:17:57 [32mBIOCONDA INFO[0m (COMMAND) conda build purge[0m
2026-04-07T14:17:57.0767359Z 
2026-04-07T14:17:57.0767363Z 
2026-04-07T14:17:57.0767367Z 
2026-04-07T14:17:58.0958318Z ⠋ running (0:00:00.00)
2026-04-07T14:17:58.0958666Z                        
2026-04-07T14:17:58.0959074Z 14:17:58 [32mBIOCONDA INFO[0m (COMMAND) docker container prune -f[0m
2026-04-07T14:17:58.0959346Z 
2026-04-07T14:17:58.0959351Z 
2026-04-07T14:17:58.0959355Z 
2026-04-07T14:17:58.1080959Z ⠋ running (0:00:00.00)
2026-04-07T14:17:58.1081322Z                        
2026-04-07T14:17:58.1082357Z 14:17:58 [31mBIOCONDA ERROR[0m BUILD SUMMARY: of 1 recipes, 1 failed and 0 were skipped. Details of recipes and environments follow.[0m
2026-04-07T14:17:58.1083540Z 14:17:58 [31mBIOCONDA ERROR[0m BUILD SUMMARY: FAILED recipe recipes/nohuman[0m
2026-04-07T14:18:04.3465480Z ##[error]Process completed with exit code 1.
```

for some more context, the ARM builds seem to work fine, it just seems to be the linux x86_64 builds. Removing stdlib('c') leads to the package being successfully built on all platforms.

--- a/bioconda_utils/__init__.py
+++ b/bioconda_utils/__init__.py
@@ -33,29 +33,7 @@ Bioconda Utilities Package
    utils
 """
 
-import re
-
-import conda_build.metadata
-
 from ._version import get_versions
-
-# Monkeypatch conda_build.metadata.find_used_variables_in_text to include __glibc
-# in the hash when stdlib('c') is used on Linux. This ensures that hashes
-# calculated with bypass_env_check=True (as done in built_package_paths)
-# match those from real builds.
-_old_find_used_variables_in_text = conda_build.metadata.find_used_variables_in_text
-
-
-def _patched_find_used_variables_in_text(variant, recipe_text, selectors_only=False):
-    used = _old_find_used_variables_in_text(variant, recipe_text, selectors_only)
-    if not selectors_only and "__glibc" in variant:
-        # Match stdlib('c')
-        if re.search(r"\{\{\s*stdlib\(\s*[\'\"]c[\'\"]\s*\)", recipe_text):
-            used.add("__glibc")
-    return used
-
-
-conda_build.metadata.find_used_variables_in_text = _patched_find_used_variables_in_text
 
 __version__ = get_versions()["version"]
 del get_versions

--- a/bioconda_utils/build.py
+++ b/bioconda_utils/build.py
@@ -493,7 +493,19 @@ def build_recipes(
             # on the host since Docker's conda-build will re-solve anyway.
             # Non-finalized metas use bypass_env_check which avoids costly
             # dependency resolution. The --no-fast-resolve flag can override this.
-            finalize = (docker_builder is None) if fast_resolve else True
+            #
+            # Two cases force finalize on to keep host and Docker hashes in sync
+            # (see https://github.com/bioconda/bioconda-utils/issues/1095):
+            #   1. Recipes using stdlib()/compiler()/pin_compatible() — their
+            #      run_exports (e.g. sysroot_linux-64 -> __glibc) are only
+            #      applied during a real solve.
+            #   2. linux-64 hosts — sysroot run_exports inject __glibc here
+            #      regardless of the recipe's text form.
+            finalize = docker_builder is None or not fast_resolve
+            if not finalize and utils.RepoData.native_platform() == "linux":
+                finalize = True
+            if not finalize and utils.recipe_requires_finalized_render(recipe):
+                finalize = True
             pkg_paths = utils.get_package_paths(
                 recipe,
                 check_channels,

--- a/bioconda_utils/utils.py
+++ b/bioconda_utils/utils.py
@@ -1090,7 +1090,7 @@ def recipe_requires_finalized_render(recipe):
     meta_path = os.path.join(recipe, "meta.yaml")
     try:
         with open(meta_path, "r", encoding="utf-8") as f:
-            text = f.read()
+            text = re.sub(r"#.*", "", f.read())
     except OSError:
         return False
     return bool(_SOLVER_DEPENDENT_JINJA.search(text))

--- a/bioconda_utils/utils.py
+++ b/bioconda_utils/utils.py
@@ -1071,6 +1071,31 @@ def changed_since_master(recipe_folder):
     ]
 
 
+# Recipe patterns whose rendered hash depends on solver state (run_exports from
+# e.g. sysroot_linux-64 inject __glibc into the variant during a real solve but
+# not under bypass_env_check=True). When any of these appear we must finalize.
+_SOLVER_DEPENDENT_JINJA = re.compile(r"\{\{\s*(stdlib|compiler|pin_compatible)\s*\(")
+
+
+def recipe_requires_finalized_render(recipe):
+    """
+    Return True if the recipe's rendered hash can depend on solver state and
+    therefore must be rendered with ``finalize=True`` to match what conda-build
+    will produce during a real build.
+
+    Detects use of ``stdlib(...)``, ``compiler(...)``, or ``pin_compatible(...)``
+    jinja functions, whose run_exports are only applied during a real solve.
+    See https://github.com/bioconda/bioconda-utils/issues/1095.
+    """
+    meta_path = os.path.join(recipe, "meta.yaml")
+    try:
+        with open(meta_path, "r", encoding="utf-8") as f:
+            text = f.read()
+    except OSError:
+        return False
+    return bool(_SOLVER_DEPENDENT_JINJA.search(text))
+
+
 def _load_platform_metas(recipe, finalize=True):
     # check if package is noarch, if so, build only on linux
     # with temp_os, we can fool the MetaData if needed.

--- a/test/test_utils.py
+++ b/test/test_utils.py
@@ -517,6 +517,10 @@ def test_recipe_requires_finalized_render(tmp_path):
     assert utils.recipe_requires_finalized_render(
         _write("requirements:\n  run:\n    - {{ pin_compatible('foo') }}\n")
     )
+    # ignore if within comments
+    assert not utils.recipe_requires_finalized_render(
+        _write("requirements:\n  run:\n    - foo  # {{ pin_compatible('foo') }}\n")
+    )
     # Missing meta.yaml -> False (don't crash)
     missing = tmp_path / "missing"
     missing.mkdir()

--- a/test/test_utils.py
+++ b/test/test_utils.py
@@ -498,6 +498,31 @@ def test_conda_as_dep(config_fixture, mulled_test):
             ensure_missing(i)
 
 
+def test_recipe_requires_finalized_render(tmp_path):
+    def _write(meta):
+        recipe = tmp_path / uuid.uuid4().hex
+        recipe.mkdir()
+        (recipe / "meta.yaml").write_text(meta)
+        return str(recipe)
+
+    assert not utils.recipe_requires_finalized_render(
+        _write("package:\n  name: pure-python\n  version: '1.0'\n")
+    )
+    assert utils.recipe_requires_finalized_render(
+        _write("requirements:\n  build:\n    - {{ stdlib('c') }}\n")
+    )
+    assert utils.recipe_requires_finalized_render(
+        _write("requirements:\n  build:\n    - {{ compiler('c') }}\n")
+    )
+    assert utils.recipe_requires_finalized_render(
+        _write("requirements:\n  run:\n    - {{ pin_compatible('foo') }}\n")
+    )
+    # Missing meta.yaml -> False (don't crash)
+    missing = tmp_path / "missing"
+    missing.mkdir()
+    assert not utils.recipe_requires_finalized_render(str(missing))
+
+
 # TODO replace the filter tests with tests for utils.get_package_paths()
 # def test_filter_recipes_no_skipping():
 #     """

